### PR TITLE
[RTM] Added ScriptHandler method to generate random secret

### PR DESCRIPTION
**Pull request by aschempp**
This adds the necessary script handler to generate the random secret for https://github.com/contao/core-bundle/issues/302

The random secret must only be generated if the `parameters.yml` file does not exist. This is because the `IncentiveeParametersHandler` will update/overwrite existing parameters if a respective environment variable is available.

Using this method means the user is not asked for a random secret on installation, but it can obviously still be changed in the `parameters.yml`.

### CODE REVIEW

**Comment by aschempp** [comment]
See https://github.com/contao/standard-edition/pull/24/files for the necessary changes in the root `composer.json` file

File: src/Composer/ScriptHandler.php [hunk 1]
```diff
@@ -42,6 +42,28 @@ public static function generateSymlinks(Event $event)
...
+            return;
+        }
+
+        $file = $extra['incenteev-parameters']['file'];
+
+        if (is_file($file)) {
+            return;
+        }
+
+        putenv('CONTAO_RANDOM_SECRET=' . md5(uniqid(mt_rand(), true)));
```
Comment (leofeyer) [inline src/Composer/ScriptHandler.php]: Is this secure? Other processes might be able to read the environment variables, too.

**Comment by discordier** [comment]
Only when executed within the same shell process and therefore belong to the same user. IMO we can assume no malicious process to be running here. If there would be, reading the environment will be the least significant problem.

**Comment by leofeyer** [comment]
>> `putenv('CONTAO_RANDOM_SECRET=' . md5(uniqid(mt_rand(), true)));`

> Is this secure? Other processes might be able to read the environment variables, too.

@ausi What's your opinion on this one?

**Comment by Toflar** [comment]
@stof Can you maybe shed some light on this? Is this the way to go or is there a security issue with this? The goal is to provide a convenience method to auto generate a secret. You can obviously still adjust it.

File: src/Composer/ScriptHandler.php (hunk 1 above)
Comment (stof) [inline src/Composer/ScriptHandler.php]: you can never change the environment of a parent process. So this environment variable will only be accessible to the current process (running Composer) and subprocesses launched from this process (subsequent composer scripts, or ``git`` commands used by the installer)

Comment (stof) [inline src/Composer/ScriptHandler.php]: not however that your implementation does not generate a strong random secret though

Comment (Toflar) [inline src/Composer/ScriptHandler.php]: Of course not, it's just convenience ;)

Comment (aschempp) [inline src/Composer/ScriptHandler.php]: @stof thanks for the feedback! Any suggestion for a stronger random secret?

Comment (leofeyer) [inline src/Composer/ScriptHandler.php]: We should probably use the Symfony security component to generate the token.

**Comment by stof** [comment]
I see a drawback here though. The Incenteev ParametersHandler considers that the environment variable contains the uptodate version of the parameter, and will replace the value coming from an existing parameters.yml file (the goal is to be able to update env variables to change the config for the next deployment, without having to delete the file).
This means that your parameter set based on your ``CONTAO_RANDOM_SECRET`` env variable will be updated for each composer run (otherwise you have a big issue with the randomness of your random secret). And this can cause many issues. For instance, changing the secret used by the remember_me token in Symfony would cause all existing remember_me tokens to become invalid (as their signature won't match using the new secret). The secret should generally not change for each deployment.

File: src/Composer/ScriptHandler.php (hunk 1 above)
Comment (stof) [inline src/Composer/ScriptHandler.php]: @aschempp ``random_bytes()`` on PHP 7, and the polyfill for PHP 5 (Symfony is deprecatign its SecureRandom class in 2.8 in favor of this new API)

**Comment by aschempp** [comment]
The variable is only set if no `parameters.yml` exists, so it will never be updated ;-)

**Comment by leofeyer** [comment]
As discussed on Mumble on October 14th, we want to use `random_bytes` with the [compatibility layer](https://github.com/paragonie/random_compat).

File: src/Composer/ScriptHandler.php (hunk 1 above)
Comment (Toflar) [inline src/Composer/ScriptHandler.php]:
```php
bin2hex(random_bytes(32));
```

Read about the polyfill but forgot about it again, thx for the hint @stof :)

Comment (stof) [inline src/Composer/ScriptHandler.php]: be careful. bin2hex does not exist in PHP 5.3

Comment (bytehead) [inline src/Composer/ScriptHandler.php]: True. The requirement from `composer.json` is anyway `"php": ">=5.4.0",`.

Comment (ausi) [inline src/Composer/ScriptHandler.php]: `bin2hex` is there since PHP 4 IMO, `hex2bin ` is 5.4 :)

**Comment by aschempp** [comment]
PR is updated

File: src/Composer/ScriptHandler.php [hunk 2]
```diff
@@ -69,4 +85,30 @@ function ($type, $buffer) use ($event) {
...
+     */
+    private static function canGenerateSecret(array $config)
+    {
+        // "incenteev-parameters" config can be an array of files.
+        if (!isset($config['file'])) {
+            $result = false;
+
+            foreach ($config as $v) {
+                if (is_array($v) && isset($v['file']) && !is_file($v['file'])) {
+                    $result = true;
```
Comment (leofeyer) [inline src/Composer/ScriptHandler.php]: Why don't you simply `return true;` here?

Comment (aschempp) [inline src/Composer/ScriptHandler.php]: Because there could be a second file that does exist (and then we must not set the variable)

Comment (leofeyer) [inline src/Composer/ScriptHandler.php]: Then you should probably return `false` if the condition matches.

**Comment by leofeyer** [comment]
Where did you enable the compat layer? You have to include the `lib/random.php` file at some point, haven't you?

File: src/Composer/ScriptHandler.php (hunk 2 above)
Comment (leofeyer) [inline src/Composer/ScriptHandler.php]: What I mean is:

```php
foreach ($config as $v) {
    if (is_array($v) && isset($v['file']) && is_file($v['file'])) {
        return false; // if there is at least one existing file, we must not generate a secret, so return false
    }
}

return true;
```

Or am I missing something here?

**Comment by stof** [comment]
> Where did you enable the compat layer? You have to include the lib/random.php file at some point, haven't you?

This is handled when loading the composer autoloader, thanks to the config of the package

File: src/Composer/ScriptHandler.php (hunk 2 above)
Comment (aschempp) [inline src/Composer/ScriptHandler.php]: You're partially right. But we also don't want to generate the variable if there is no file defined in the array, where in your example we would return `true`

Comment (leofeyer) [inline src/Composer/ScriptHandler.php]: Ok, thanks for checking. The PR is RTM to me then.

**Comment by aschempp** [comment]
Ok, I have updated the file check again based on @leofeyer's input and added full unit tests for the new method.

**Comment by leofeyer** [comment]
Merged in 2c2fea1dad449c9b5f59b92b90b45f30dcbee505.

**Pull request merged**